### PR TITLE
Reduce the number of terminals required to build riot-web to 1

### PR DESCRIPTION
**Pull request by turt2live**
~~**This is a work in progress.**~~

todo:
* [x] Fix race condition where the react-sdk initial build fails sometimes
* [x] Fix race condition where the riot-web initial build fails sometimes
* [x] Fix race condition where riot-web builds 5+ times during the initial build

----

A step towards a real solution for https://github.com/vector-im/riot-web/issues/7305

This approach makes use of `npm link` to remove the use of symlinks in the build process. The build process has also been altered to invoke the build process of each underlying SDK (react, js). This means that one can now `npm link` and `npm start` and have a working environment. 

At the same time, parallelshell was dropped due to lack of maintenance from the maintainer.

Requires:
* [x] https://github.com/matrix-org/matrix-js-sdk/pull/742
* [x] https://github.com/matrix-org/matrix-react-sdk/pull/2175

### CODE REVIEW

**Comment by turt2live** [comment]
Race conditions addressed 🎉

File: scripts/block-on-sdk-build.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        let timerId = null;
+
+        const watcher = chokidar.watch(buildDirectory).on('all', (event, path) => {
+            lock.acquire("timer", (done) => {
+                if (timerId !== null) {
+                    //console.log("Resetting countdown");
+                    clearTimeout(timerId);
+                }
+                //console.log(`Waiting ${WAIT_TIME}ms for another file update...`);
+                timerId = setTimeout(() => {
```
Comment (ara4n) [inline scripts/block-on-sdk-build.js]: Hm, i'm surprised that we pause 5s at this point, and i'm failing to see why...

**Comment by ara4n** [comment]
this generally lgtm, although i don't understand why we need the 5s timeout given we're doing the swanky (dying) canary thing.

also, the comment in https://github.com/vector-im/riot-web/commit/2b037ee146dcfe6ed40f5848ecd825e76c58fa3d would really benefit from going in a comment block somewhere, perhaps at the top of block-on-sdk-build.sh?

exciting!

**Comment by turt2live** [comment]
I'll reword and put the comment block in somewhere.

We do a 5s timeout for file changes because the canary only indicates when we should actually start watching for file changes. The high level process looks like `npm install (causes a build) -> canary -> start watching for changes -> npm start` for an SDK. The 5s timeout is there to give the `npm start` (webpack/babel) enough time to process larger files and ensure the process is actually finished. In my testing, this could go as low as 1 or 1.5s, however that caused some instability. 5s was arbitrarily chosen as "safe", and means the initial build is only delayed by ~10s.

**Comment by ara4n** [comment]
why can't we fire the canary once npm start has stopped thrashing?

**Comment by turt2live** [comment]
Because babel (used by the js-sdk) doesn't provide a good place to do this, and webpack (used by the react-sdk) doesn't seem to want to behave and do that. `block-on-sdk-build.js` effectively does the job though by waiting for the thrashing to stop before letting the process continue, and works with any build system.

**Comment by ara4n** [comment]
right, i see. i guess one alternative could be to chuck a dummy file called zz9.js into js-sdk/src and rely on the fact that babel traverses the FS in alphabetical order, and use js-sdk/lib as a canary to tell once babel has finished its thing? this looks like it could also work in react-sdk, which seems to be using babel rather than webpack?

My concern about the 5s is mainly the risk of slower/overloaded comps or CI systems taking longer and a general aversion to 'sticking a sleep in it', as well as wanting first experiences to be as snappy as possible.

**Comment by turt2live** [comment]
CI should be unaffected as this doesn't touch that path, however 5s is fairly arbitrary indeed. Relying on alphabetical ordering of babel feels slightly more magical than a sleep to me, and doesn't really give the code base a good image (imo) when there's a magic file included in the source tree.

**Comment by turt2live** [comment]
That sounds like a safer option (and also uses a lot less magic watching). I'll give that a shot tomorrow and see how it goes, expecting it to be snappier.

**Comment by ara4n** [comment]
lgtm modulo the thoughts you already filed over at https://github.com/vector-im/riot-web/issues/7386, assuming it works :)

**Comment by ara4n** [comment]
lgtmer. (i have slight concerns about what `npm link` is buying us here; but i guess if the user wishes to manually symlink rather than using `npm link` they can)

**Comment by ara4n** [comment]
merging so i can use it :P

**Pull request merged**